x-pack/plugins/maps/server/index_data.ts

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IScopedClusterClient } from 'kibana/server';
+import { Logger } from 'src/core/server';
+import {
+  INDEX_META_DATA_CREATED_BY,
+  ImportResponse,
+  ImportFailure,
+  Settings,
+  Mappings,
+} from '../common';
+import { IndexPatternsService } from '../../../../src/plugins/data/common';
+
+export type InputData = any[];
+interface BodySettings {
+  [key: string]: any;
+}
+
+const DEFAULT_SETTINGS = { number_of_shards: 1 };
+const DEFAULT_MAPPINGS = {
+  _meta: {
+    created_by: INDEX_META_DATA_CREATED_BY,
+  },
+};
+
+export function indexDataProvider(
+  { asCurrentUser }: IScopedClusterClient,
+  indexPatternsService: IndexPatternsService,
+  logger: Logger
+) {
+  async function indexData(
+    index: string,
+    mappings: Mappings,
+    data: InputData
+  ): Promise<ImportResponse> {
+    const docCount = data.length;
+    enum IndexStep {
+      CREATING_INDEX = 'creating index',
+      CREATING_INDEX_PATTERN = 'creating index pattern',
+      INDEXING_DATA = 'indexing data',
+    }
+    let currentStep: IndexStep = IndexStep.CREATING_INDEX;
+
+    try {
+      await createIndex(index, mappings);
+      currentStep = IndexStep.CREATING_INDEX_PATTERN;
+      await createIndexPattern(index);
+
+      let failures: ImportFailure[] = [];
+      currentStep = IndexStep.INDEXING_DATA;
+      const resp = await writeDataToIndex(index, data);
+      if (resp.success === false) {
+        if (resp.ingestError) {
+          throw resp;
+        } else {
+          logger.warn(`Error: some documents failed to index. ${resp.failures}`);
+          failures = resp.failures || [];
+        }
+      }
+
+      return {
```

Comment:
Lets make the return type simpler. The method should just return success or failure with may be an error, no need to copy file upload response shape.